src/main/java/com/shair13/external_service/client/MovieRestClientImpl.java

```diff
@@ -0,0 +1,103 @@
+                .onStatus(HttpStatusCode::is5xxServerError, clientResponse -> {
+                    throw new InternalServerErrorException("Server Error");
+                })
+                .bodyToMono(Movie.class)
+                .block();
+    }
+
+    @Override
+    public Movie getById(Long id) {
+        return webClient.get()
+                .uri("/movies/{id}", id)
+                .retrieve()
+                .onStatus(HttpStatusCode::is4xxClientError, clientResponse -> {
+                    throw new MovieNotFoundException(id);
+                })
+                .onStatus(HttpStatusCode::is5xxServerError, clientResponse -> {
+                    throw new InternalServerErrorException("Server Error");
+                })
+                .bodyToMono(Movie.class)
+                .block();
+    }
+
+    @Override
+    public PagedMovie search(MovieSearchParams searchParams, PageDetails pageParams) {
+        return webClient.get()
+                .uri(uriBuilder -> uriBuilder
+                        .path("/movies")
+                        .queryParam("page", pageParams.getPage())
+                        .queryParam("size", pageParams.getSize())
+                        .queryParam("sort-by", pageParams.getSortBy())
+                        .queryParam("title", searchParams.getTitle())
+                        .queryParam("director", searchParams.getDirector())
+                        .queryParam("description", searchParams.getDescription())
+                        .queryParam("rate-gt", searchParams.getRateGreaterThan())
+                        .build())
+                .retrieve()
+                .onStatus(HttpStatusCode::is5xxServerError, clientResponse -> {
+                    throw new InternalServerErrorException("Server Error");
+                })
+                .onStatus(HttpStatusCode::is4xxClientError, clientResponse -> {
+                    throw new InternalServerErrorException("Bad Request");
+                })
+                .bodyToMono(PagedMovie.class)
+                .block();
+    }
+
+    @Override
+    public Movie update(Long id, Movie movie) {
+        return webClient.put()
+                .uri("/movies/{id}", id)
+                .body(Mono.just(movie), Movie.class)
+                .retrieve()
+                .onStatus(HttpStatusCode::is4xxClientError, clientResponse -> {
+                    throw new MovieNotFoundException(id);
+                })
+                .onStatus(HttpStatusCode::is5xxServerError, clientResponse -> {
+                    throw new InternalServerErrorException("Server Error");
+                })
+                .bodyToMono(Movie.class)
+                .block();
+    }
+
+    @Override
+    public void delete(Long id) {
+        webClient.delete()
+                .uri("/movies/{id}", id)
+                .retrieve()
+                .onStatus(HttpStatusCode::is4xxClientError, clientResponse -> {
+                    throw new MovieNotFoundException(id);
```

Comment:
Lepiej byłoby rzucać MovieNotFoundException tylko w przypadku wystąpienia błędu 404. Pomimo braku jawnej obsługi innych błędów, mogą one się pojawić - np. w przypadku przekazania niewłaściwego formatu danych (spring to obsłuży i wyrzuci np. 400 Bad Request).

Reply:
Zmieniłem rzucanie wyjątu MovieNotFoundException tylko w sytuacji wystąpienia błędu 404. Dodatkowo usunąłem błąd 400 Bad Request ponieważ faktycznie spring sam go przechwytuje i wyrzuca.